driver/src/main/com/mongodb/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -0,0 +1,159 @@
+    private final int chunkSizeInBytes;
+    private final int numberOfChunks;
+    private int chunkIndex;
+    private int bufferOffset;
+    private long currentPosition;
+    private byte[] buffer = null;
+    private boolean closed;
+
+    GridFSDownloadStreamImpl(final GridFSFile fileInfo, final MongoCollection<BsonDocument> chunksCollection) {
+        this.fileInfo = notNull("file information", fileInfo);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+
+        fileId = fileInfo.getId();
+        length = fileInfo.getLength();
+        chunkSizeInBytes = fileInfo.getChunkSize();
+        numberOfChunks = (int) Math.ceil((double) length / chunkSizeInBytes);
+    }
+
+    @Override
+    public GridFSFile getGridFSFile() {
+        return fileInfo;
+    }
+
+    @Override
+    public int read() throws IOException {
+        byte[] b = new byte[1];
+        int res = read(b);
+        if (res < 0) {
+            return -1;
+        }
+        return b[0] & 0xFF;
+    }
+
+    @Override
+    public int read(final byte[] b, final int off, final int len) throws IOException {
+        checkClosed();
+
+        if (currentPosition == length) {
+            return -1;
+        } else if (buffer == null) {
+            buffer = getBuffer(chunkIndex);
+            bufferOffset = 0;
+        } else if (bufferOffset == buffer.length) {
+            chunkIndex += 1;
+            buffer = getBuffer(chunkIndex);
+            bufferOffset = 0;
+        }
+
+        int r = Math.min(len, buffer.length - bufferOffset);
+        System.arraycopy(buffer, bufferOffset, b, off, r);
+        bufferOffset += r;
+        currentPosition += r;
+        return r;
+    }
+
+    @Override
+    public long skip(final long bytesToSkip) throws IOException {
+        checkClosed();
+        if (bytesToSkip <= 0) {
+            return 0;
+        }
+
+        long skippedPosition = currentPosition + bytesToSkip;
+        bufferOffset = (int) skippedPosition % chunkSizeInBytes;
+        if (skippedPosition >= length) {
+            long skipped = length - currentPosition;
+            chunkIndex = numberOfChunks - 1;
+            currentPosition = length;
+            buffer = null;
+            return skipped;
+        } else {
+            chunkIndex = (int) Math.floor((float) skippedPosition / chunkSizeInBytes);
+            currentPosition += bytesToSkip;
+            buffer = getBuffer(chunkIndex);
+            return bytesToSkip;
+        }
+    }
+
+    @Override
+    public int available() throws IOException {
+        checkClosed();
+        if (buffer == null) {
+            return 0;
+        } else {
+            return buffer.length - bufferOffset;
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        checkClosed();
+        closed = true;
+    }
+
+    private void checkClosed() {
+        if (closed) {
+            throw new MongoGridFSException("The InputStream has been closed");
```

Comment:
After speaking because other operations could throw a MongoException keeping it as it is.

Done.